internal/clonepool/clonepool.go

```diff
@@ -0,0 +1,245 @@
+package clonepool
+
+import (
+	"fmt"
+	"github.com/labstack/gommon/log"
+	"github.com/modfin/henry/slicez"
+	"github.com/modfin/zdap"
+	"github.com/modfin/zdap/internal"
+	"github.com/modfin/zdap/internal/cloning"
+	"github.com/modfin/zdap/internal/zfs"
+	"sync"
+	"time"
+)
+
+type ClonePool struct {
+	resource        internal.Resource
+	cloneContext    *cloning.CloneContext
+	ClonesAvailable int
+	claimLock       sync.Mutex
+}
+
+func NewClonePool(resource internal.Resource, cloneContext *cloning.CloneContext) ClonePool {
+	return ClonePool{resource: resource, cloneContext: cloneContext}
+}
+
+func (c *ClonePool) Start() {
+	go func() {
+		for {
+			time.Sleep(time.Second)
+			c.action()
+		}
+	}()
+}
+
+func (c *ClonePool) action() {
+
+	dss, err := c.cloneContext.Z.Open()
+	defer dss.Close()
+	if err != nil {
+		fmt.Printf("error trying to open z, error: %s\n", err.Error())
+		return
+	}
+
+	allClones, err := c.readPooled(dss)
+	if err != nil {
+		fmt.Printf("could not read pooled clones, error %s", err)
+		return
+	}
+
+	if c.resource.ClonePool.ResetOnNewSnap {
+		err = c.expireClonesFromOldSnaps(dss)
+		if err != nil {
+			fmt.Printf("could not expire old clones, error %s", err)
+		}
+	}
+
+	nonExpiredClones := c.pruneExpired(dss, allClones)
```

Comment:
Right. We could extract a new `Expire` method here that accepts a claimId **and** a dataset, so that the entire c.action works on the same opened dataset.

Reply:
I may have been wrong about this, I checked the `GetUserProperty` and `SetUserProperty` and it seems as if those isn't cached in memory. So this will probably work as intended. But good at least to have `expire` that can use the dataset that we already have loaded when `action` is triggered.